### DISCUSSION TITLE
Typescript performance in comparison to MST

### DISCUSSION THREAD

**Discussion by evelant** [xaviergonz/mobx-keystone#272] [Q&A]
How does the Typescript performance during editing compare between mobx-keystone and mobx-state-tree?

I have an app with a lot of large MST models and this has greatly slowed down TS editing performance to the point where it impacts productivity. TSServer can't provide completions faster than 5+ seconds anymore which is a hindrance when working with models that have hundreds of properties. I _think_ this is due to the design of the type definitions for mobx-state-tree having a high complexity and relying on a lot of inference from TS.

Is typescript editor performance better with large models when using mobx-keystone?

**Comment by xaviergonz** [comment]
in theory it should be, but I don't have any benchmarks to prove it

**Comment by mindreframer** [comment]
@AndrewMorsillo the performance is better, I also have started with Mobx State Tree and the editing in VSC became quite sluggish, now with Mobx Keystone it is quite snappy. I would recommend creating a small empty project and re-implement those models there with Keystone. This should give you a good feeling about how it would feel later.

**Comment by evelant** [comment]
@mindreframer Thanks for letting me know it was an improvement for you. I need to think through how I might start migrating from MST to mobx-keystone. Problem is I have a _lot_ of big MST models which all depend on each other, it would be a huge job to try to refactor it all at once.

**Comment by mindreframer** [comment]
@AndrewMorsillo  Thats right. You should do it like bridge-builders do it: they do not destroy the bridge, but first repair one side, then the other, while letting the bridge still be usable. 

I would suggest creating a new folder for Keystone models, that is completely unused by your UI and try to re-implement the store API within that folder. To avoid huge pull requests, merge it often to your main / master branch, it should not break anything and live along your MST models. If you encounter some challenges, you can also adjust the MST API, to make the later "swapping" smoother. Copy-paste is OK. That way you avoid blocking all other progress, see what kind of issues you encounter along the way and keep your sanity. You will know, that at any moment you would be able to abort this experiment and return back to MST models (working, yet problematic in dev).
